problem/views.py

```diff
@@ -30,24 +41,86 @@ def problem(request):
-      'description': 'Given a, b, output a+b.',
-      'input': 'a, b <= 100000000',
-      'output': 'a+b',
-      'samp_input': '1 2\n4 5',
-      'samp_output': '3\n9\n',
-      'tag': [''],
-      'testcase': [{'num': 1, 'time': 1, 'memory': 32}, {'num': 2, 'time': 3, 'memory': 100}],
-    }
-    return render(request, 'problem/detail.html', p)
+    logger.info('detail of problem %s' % (problem_id))
+    problem = Problem.objects.get(pk=problem_id)
+    testcase = Testcase.objects.filter(problem=problem)
+    tag = problem.tags.all()
+    return render(request, 'problem/detail.html', 
+                  { 'problem': problem, 'tag': tag, 'testcase': testcase })
 
 def edit(request, problem_id):
-    return render(request, 'problem/edit.html')
+    logger.info('edit problem %s' % (problem_id))
+    problem = Problem.objects.get(pk=problem_id)
+    testcase = Testcase.objects.filter(problem=problem)
+    tags = problem.tags.all()
+    if request.method == 'GET':
+        form = ProblemForm(instance=problem)
+    if request.method == 'POST':
+        form = ProblemForm(request.POST, instance=problem)
+        if form.is_valid():
+            problem = form.save()
+            problem.description = request.POST['description']
+            problem.input= request.POST['input_description']
+            problem.output = request.POST['output_description']
+            problem.sample_in = request.POST['sample_input']
+            problem.sample_out = request.POST['sample_output']
+            problem.save()
+            return redirect('/problem/%d' % (problem.pk))
+    return render(request, 'problem/edit.html', 
+                  { 'form': form, 'pid': problem_id, 'is_new': False, 'tags': tags,
+                   'description': problem.description,
+                   'input': problem.input, 'output': problem.output,
+                   'sample_in': problem.sample_in, 'sample_out': problem.sample_out,
+                   'testcase': testcase })
 
 def new(request):
-    return render(request, 'problem/edit.html')
+    if request.method == 'GET':
+        form = ProblemForm()
+    if request.method == 'POST':
+        logger.info('post new problem')
+        form = ProblemForm(request.POST)
+        if form.is_valid():
+            problem = form.save()
+            problem.description = request.POST['description']
+            problem.input= request.POST['input_description']
+            problem.output = request.POST['output_description']
+            problem.sample_in = request.POST['sample_input']
+            problem.sample_out = request.POST['sample_output']
+            problem.save()
+            return redirect('/problem/%d' % (problem.pk))
+    return render(request, 'problem/edit.html', { 'form': form, 'is_new': True })
+
+def tag(request, problem_id):
+    if request.method == 'POST':
+        tag = request.POST['tag']
+        problem = Problem.objects.get(pk=problem_id)
+        print problem.tags.filter(tag_name=tag).exists()
```

Comment:
Don't use print directly. You may return an error message and show it on web page.